PG16 : Updates to make_restrictinfo

lkshminarayanan opened the pull request:
[Remove unused functions](https://github.com/timescale/timescaledb/pull/5979/commits/2127d85e8a8ad95abd77d679b28aadc2262fbfcd) 

Removed the unused functions ts_make_pathkey_from_sortop and
ts_make_pathkey_from_sortinfo.

Note : Patch unrelated to PG16 but these functions call get_eclass_for_sort_expr that is affected by PG16 changes.

---

[PG16: Handle removed EquivalenceClass members](https://github.com/timescale/timescaledb/commit/8791b3bfb47585e538a1be0402aaf81643743668) 

PG16 removes the outerjoin_delayed mechanism and
RestrictInfo.nullable_relids.

https://github.com/postgres/postgres/commit/b448f1c8
https://github.com/postgres/postgres/commit/3bef56e1

---

[PG16: Handle updates to make_restrictinfo function](https://github.com/timescale/timescaledb/commit/ad1d0a32d110b36e49db3404cd8d578fe89c94d2) 

While fixing a bug in filtering of "cloned" outer-join quals, PG16 adds
3 new parameters to the make_restrictinfo function. Updated the compat
function to handle this change. This patch also cleans up the variants
of make_restrictinfo from other versions to make it clear which args are
passed.

https://github.com/postgres/postgres/commit/991a3df227

---

Disable-check: force-changelog-file
Disable-check: commit-count

mkindahl commented on tsl/src/fdw/data_node_scan_plan.c:
```diff
@@ -201,11 +201,19 @@ adjust_data_node_rel_attrs(PlannerInfo *root, RelOptInfo *data_node_rel, RelOptI
 								make_restrictinfo_compat(root,
 														 (Expr *) onecq,
 														 rinfo->is_pushed_down,
+#if PG16_LT
+														 false,
+														 false,
+#else
+														 rinfo->has_clone,
+														 rinfo->is_clone,
+#endif
 														 rinfo->outerjoin_delayed,
```
It is, in general, not a good idea to have preprocessor directives inside the parameter list of a preprocessor macro. Weird issues can result. See https://wiki.sei.cmu.edu/confluence/display/c/PRE32-C.+Do+not+use+preprocessor+directives+in+invocations+of+function-like+macros

mkindahl commented on src/sort_transform.c:
```diff
@@ -272,7 +272,9 @@ sort_transform_ec(PlannerInfo *root, EquivalenceClass *orig)
 			 */
 			EquivalenceClass *exist = get_eclass_for_sort_expr(root,
 															   transformed_expr,
+#if PG16_LT
 															   ec_mem->em_nullable_relids,
+#endif
 															   opfamilies,
 															   type_oid,
 															   orig->ec_collation,
```
Even though this works since `get_eclass_for_sort_expr` is not a macro, I would avoid preprocessor macros inside parameter lists.

mkindahl commented on tsl/src/nodes/decompress_chunk/decompress_chunk.c:
```diff
@@ -1399,25 +1404,29 @@ add_segmentby_to_equivalence_class(EquivalenceClass *cur_ec, CompressionInfo *in
 		new_relids = bms_add_members(new_relids, info->compressed_rel->relids);
 
+#if PG16_LT
 		/*
 		 * And likewise for nullable_relids.  Note this code assumes
 		 * parent and child relids are singletons.
 		 */
-		new_nullable_relids = cur_em->em_nullable_relids;
+		Relids new_nullable_relids = cur_em->em_nullable_relids;
 		if (bms_overlap(new_nullable_relids, uncompressed_chunk_relids))
 		{
 			new_nullable_relids = bms_difference(new_nullable_relids, uncompressed_chunk_relids);
 			new_nullable_relids =
 				bms_add_members(new_nullable_relids, info->compressed_rel->relids);
 		}
+#endif
```
Nit: Personally, I would put this in a block just to make sure that `new_nullable_relids` do not "leak out" to the following code.

lkshminarayanan replied:
The new_nullable_relids is actually used in a block that follows this. I have now moved this code together with it and both protected by a PG16_LT macro.

mkindahl reviewed (approved):
Looks safe, so approving, but this usage of pre-processor macros to comment out parameters can be problematic.

lkshminarayanan commented:
> Looks safe, so approving, but this usage of pre-processor macros to comment out parameters can be problematic.

I have now updated the patch to remove such usages. I also added a new commit that removes an unused function.
